Iterative implementation for newickize

ttung opened the pull request:
`newickize` can [overflow the python stack for very deep trees](https://github.com/nextstrain/augur/issues/328).  This is an alternate implementation that uses a deque (as a stack) to store the tree data.  Although this is the same O(N) in terms of storage, the amount of data stored per function call depth in python is significantly higher.

<!--- Please read each of the following items and confirm by replacing
 !--the [ ] with a [X] --->

- [x] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [x] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [x] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

ttung commented:
I am checking with my employer if there are any issues with me ticking off the licensing question.

peterjc commented:
Thanks for checking about licensing, some employers can be quite strict about this.

Note you will need to run ``black`` on the code, which will fix some of the style issues failing in the ``flake8`` checks:

```
$ flake8
./Bio/Phylo/NewickIO.py:291:1: D205 1 blank line required between summary line and description
./Bio/Phylo/NewickIO.py:291:1: D209 Multi-line docstring closing quotes should be on a separate line
./Bio/Phylo/NewickIO.py:291:1: D400 First line should end with a period
./Bio/Phylo/NewickIO.py:298:1: BLK100 Black would make changes.
./Bio/Phylo/NewickIO.py:299:1: D204 1 blank line required after class docstring
./Bio/Phylo/NewickIO.py:299:1: D205 1 blank line required between summary line and description
./Bio/Phylo/NewickIO.py:299:1: D209 Multi-line docstring closing quotes should be on a separate line
./Bio/Phylo/NewickIO.py:299:1: D400 First line should end with a period
./Bio/Phylo/NewickIO.py:305:1: D204 1 blank line required after class docstring
./Bio/Phylo/NewickIO.py:305:1: D205 1 blank line required between summary line and description
./Bio/Phylo/NewickIO.py:305:1: D209 Multi-line docstring closing quotes should be on a separate line
./Bio/Phylo/NewickIO.py:305:1: D400 First line should end with a period
```

peterjc commented on Bio/Phylo/NewickIO.py:
```diff
@@ -287,18 +288,60 @@ def to_strings(
-            else:
-                subtrees = (newickize(sub) for sub in clade)
-                return "(%s)%s" % (",".join(subtrees), label + make_info_string(clade))
+            """Convert a node tree to a Newick tree string, iteratively.  This is done
```
Put a blank line after "..., iteratively." to follow the PEP257 docstring convention. This is important for extracting the docstrings for API documentation.

ttung replied:
Fixed!

ttung commented:
> Note you will need to run black on the code, which will fix some of the style issues failing in the flake8 checks

Done!

ttung commented:
As far as I can tell, none of the travis failures are related to my changes.  If I am incorrect, some guidance would be helpful.

Also, is PR review contingent on licensing checkmark?  I would like to know how hard to push the legal folks. :)

peterjc commented:
Merging a pull request is contingent on the licensing agreement.

TravisCI is failing to build the API documentation, but I saw that same error earlier on another pull request, so it is unrelated. Probably a version change in Sphinx or something like that... see https://github.com/biopython/biopython/pull/3021#issuecomment-652927061

ttung commented:
> Merging a pull request is contingent on the licensing agreement.

Cool, just wanted to make sure legal wasn't blocking the code review. :)

ttung commented:
I've gotten approval from legal to check the box.  :)

Looking forward to getting this merged.

Commit:
Iterative implementation for newickize

`newickize` can [overflow the python stack for very deep trees](https://github.com/nextstrain/augur/issues/328).  This is an alternate implementation that uses a deque (as a stack) to store the tree data.  Although this is the same O(N) in terms of storage, the amount of data stored per function call depth in python is significantly higher.

tsibley commented:
Just wanted to chime in to say that I'm interested in seeing this merged and released. :-)

The [AppVeyor CI failure](https://ci.appveyor.com/project/biopython/biopython/builds/34069230/job/xu5l3yhxma08d5rq#L1597) seems to be unrelated at first glance (and based on the error message, possibly transient?).

peterjc commented:
Yes, ``psycopg2.OperationalError`` issue logged as #3070,

```
======================================================================
ERROR: test_protein_refseq2 (common_BioSQL.TransferTest)
From GenBank file to BioSQL, then again to a new namespace, protein_refseq2.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\biopython\Tests\common_BioSQL.py", line 878, in setUp
    TESTDB = create_database()
  File "C:\projects\biopython\Tests\common_BioSQL.py", line 220, in create_database
    driver=DBDRIVER, user=DBUSER, passwd=DBPASSWD, host=DBHOST, db=TESTDB
  File "C:\projects\biopython\build\lib.win32-3.7\BioSQL\BioSeqDatabase.py", line 110, in open_database
    conn = connect(**kw)
  File "C:\Py\lib\site-packages\psycopg2\__init__.py", line 127, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
psycopg2.OperationalError: server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
```

Will rerun that now...

peterjc commented:
@etal are you able to review this, or suggest someone else who could?

ttung commented:
Any update on getting this in? :)

ttung commented:
Ping! :)
